xwayland/xwm.c

```diff
@@ -540,26 +598,27 @@ static void handle_client_message(struct wlr_xwm *xwm,
 	wlr_log(L_DEBUG, "XCB_CLIENT_MESSAGE (%u)", ev->window);
 
 	if (ev->type == xwm->atoms[WL_SURFACE_ID]) {
-		struct wlr_xwayland_surface *surface = lookup_surface(
-			&xwm->new_surfaces, ev->window);
+		struct wlr_xwayland_surface *surface = lookup_surface(xwm, ev->window);
```

Comment:
Well, my naming sense sucks but the way the code was layout with three lists really had a purpose :)

So now I have the code under my nose, what the old code did was (I want a blackboard, parallelism sucks to explain in text):
 * wlr_xwayland_surface_create is called for the X11 side (XCB_CREATE_NOTIFY event), and creates the surface / puts it in new_surfaces
 * At some point later, X11 sends XCB_CLIENT_MESSAGE with type WL_SURFACE_ID. This is meant to let us map the surface that was created through wayland with that ID and the X11 window with that other ID: the only possible candidates are windows which got created but never got this message yet, hence a list just for that for quick lookup.
At this point either the wayland surface exist and we map it, or it's not there yet and we put the wlr_xwayland_surface in the unpaired list to remember the surface_id and look at it when the wayland surface is created.
 * At some point in parallel (before or after any of the other two bullets), wayland creates the surface. Either it's in the unpaired list (both other events were first) and we can map it, or we're first and the WL_SURFACE_ID message will come later and find this surface through wl_client_get_object so we have nothing to do.

To sum up, during the life of a wlr_xwayland_surface it would either be put in new_surfaces then displayable_surfaces if the wayland event was first;  or new_surfaces then unpaired_surfaces then displayable_surfaces if the xcb event was first.

Now, this change will technically work but this bit bothers me slightly. This lookup here will go through all the X11 windows, when there really only are very few candidate (this list will usually only have one element, so that particular lookup was almost free and is now going through everything). I don't suppose most people will have thousands of X11 clients around so the performance impact is probably fine, but this risks remapping windows that already had been mapped if X11 goes crazy -- I'm hardly convinced this code will work in that case and I do not trust Xwayland (I have no clue if and X11 client could make it send arbitrary messages) so this is a DoS or worse waiting to happen.
I guess the code is simpler that way and it might be worth it, but we need to make sure things work in formerly-unexpected cases.

(I definitely agree with getting away from the displayable_windows list that was public -- I was using that list in the original compositor because there was nothing handling the mapping events, but compositors will be managing this side on their own. Unfortunately we still need a list/hash/whatever with all the windows because xcb talks to us with only windows ID and we need to find which structure corresponds to that)

Reply:
> but this risks remapping windows that already had been mapped if X11 goes crazy

Can you go through this and add assertions so we crash gracefully and get good bug reports?

Reply:
No clue how to add commits to an existing PR (I could do a PR to your branch on your repo but what a mess that is to track), I'll wait till this gets merged and have another look at this.